app/scripts/metamask-controller.js

```diff
@@ -1506,6 +1508,50 @@ module.exports = class MetamaskController extends EventEmitter {
     )
   }
 
+  onMessage (message, sender, sendResponse) {
+    if (!message || !message.type) {
+      log.debug(`Ignoring invalid message: '${JSON.stringify(message)}'`)
+      return
+    }
+
+    let handleMessage
+
+    try {
+      if (message.type === 'metamask:registerOnboarding') {
+        assert(sender.tab, 'Missing tab from sender')
+        assert(sender.tab.id && sender.tab.id !== extension.tabs.TAB_ID_NONE, 'Missing tab ID from sender')
+        assert(message.location, 'Missing location from message')
+
+        handleMessage = this.onboardingController.registerOnboarding(message.location, sender.tab.id)
+      } else if (message.type === 'metamask:registerOnboardingForwarder') {
+        assert(message.openerLocation, 'Missing opener location')
+        // TODO: filter by forwarder domain unless in testing/development
```

Comment:
This remains to be done. The forwarder needs to be deployed first for this to work though.